### PR TITLE
Add citation and references

### PR DESCRIPTION
**Pull request by rollaball**
* **Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Adds Components for Citation and References.

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**
Add [Cite/] anywhere on the page with props authors, title and source.
Then Add [Cite.References /] and it will get the list of all citations on the page

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Other information**:
Issue: #509

### CODE REVIEW

**Comment by rollaball** [comment]
```
Idyll is a markup language and toolkit for writing interactive articles. [Cite 
  source:"https://idl.cs.washington.edu/files/2018-Idyll-UIST.pdf" 
  authors:"Matthew Conlen and Jeffrey Heer"
  title:"Idyll: A Markup Language for Authoring and Publishing Interactive Articles on the Web"
/]
![Screenshot from 2019-07-31 15-13-36](https://user-images.githubusercontent.com/14182492/62201851-cd7ce180-b3a5-11e9-9697-e594f90fecbd.png)


[Cite.References/]

```

**Comment by mathisonian** [comment]
This looks great @rollaball! 

One use case this doesn't quite cover yet is a way to cite via an id. For example it is common to cite the same work multiple times, it would be nice if there were a way to provide a shortcut if you want to reference something again, without typing in all the information. Can you think of a clean way to do that?

**Comment by rollaball** [comment]
```[Cite author:"..." title:"..." url:"..." referenceName:"..." /]
[Cite referenceName:"..." /]```
@mathisonian does this seem like a good way to do this?

**Comment by mathisonian** [comment]
That works for me! Only comment is that `referenceName` is a little verbose, what about just `name` or `id`? Of those two `id` maybe is clearer than name since one might reasonable think `name` is referring to the author name or paper title. @fredhohman you have any comment on this?

We should probably print a warning if you provide a `referenceName` but it doesn't have any associated data yet. 

I like this API because it doesn't seem like it would need any changes if we also add bibtex support in the future.

File: packages/idyll-components/src/cite.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    referenceInstances.push(this);
+  }
+  render() {
+    return (
+      <div id="references">
+        <h1>References</h1>
+        <ol>
+          {citations.map((citation, index) => (
+            <li key={index}>
+              {`${citation.authors}: `}
```
Comment (mathisonian) [inline packages/idyll-components/src/cite.js]: Can this be simplified to `{citation.authors}:` or is this a whitespace preservation thing?

**Comment by fredhohman** [comment]
Nice!

I agree `id` is more clear, because `name` could be referring to a number of different fields of a citation/reference/publication.

Other examples:
* I think BibTeX calls it a "citation-key": http://www.bibtex.org/Format/
* Distill calls it a `key` as well. Here's link from their example article template with multiple citations as well: https://github.com/distillpub/template/blob/master/examples/article-test.html#L95

**Comment by mathisonian** [comment]
It should probably accept a `date` property as well that takes an arbitrary string.

**Comment by mathisonian** [comment]
Thanks for the links @fredhohman! 

@rollaball would calling it `key` mess with react's special `key` property?

**Comment by rollaball** [comment]
@mathisonian 
Tried it and it is messing with the special `key`. Will have to call it something else.

**Comment by rollaball** [comment]
@mathisonian 
so what should I go ahead with? `id` or `citation-key`?

**Comment by mathisonian** [comment]
@rollaball let's go with `id`

File: packages/idyll-components/src/cite.js (hunk 1 above)
Comment (rollaball) [reply packages/idyll-components/src/cite.js]: Replacing it with this to preserve whitespace
```
 {citation.authors}:{' '}
     <a href={citation.url} target="_blank">
```
Or should i remove it all together?

Comment (mathisonian) [reply packages/idyll-components/src/cite.js]: That works 👍

**Comment by rollaball** [comment]
@mathisonian 
I have made the changes

File: packages/idyll-components/src/cite.js (hunk 1 above)
Comment (rollaball) [reply packages/idyll-components/src/cite.js]: Done

File: packages/idyll-components/src/cite.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      console.warn(
+        `Warning: Cite with given id not found or invalid id. Check if a Cite with the id "${id}" was declared before this line.`
+      );
+      this.state = {};
+    }
+  }
+  render() {
+    const { citationNumber, authors, title } = this.state;
+    return (
+      <a title={`${authors}: ${title}`} href="#references">
```
Comment (mathisonian) [inline packages/idyll-components/src/cite.js]: Can we hyperlink to the specific citation number?

File: packages/idyll-components/src/cite.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        <h1>References</h1>
+        <ol>
+          {citations.map((citation, index) => (
+            <li key={index}>
+              {citation.authors}:{' '}
+              <a href={citation.url} target="_blank">
+                {citation.title}
+              </a>
+              , {citation.date}
+            </li>
```
Comment (mathisonian) [inline packages/idyll-components/src/cite.js]: Let's change the format to:

---

[Title with optional link](https://idyll-lang.org), Author Names. Optional Venue. Optional Date. 

---

It should gracefully degrade if properties aren't provided, so e.g. if no date is provided there shouldn't be a double period at the end.

Comment (mathisonian) [reply packages/idyll-components/src/cite.js]: Just made a small edit in the comment above 👆

File: packages/idyll-components/src/cite.js (hunk 2 above)
Comment (rollaball) [reply packages/idyll-components/src/cite.js]: updated

File: packages/idyll-components/src/cite.js (hunk 3 above)
Comment (rollaball) [reply packages/idyll-components/src/cite.js]: After change:
Screenshot from 2019-08-04 03-59-16

**Comment by mathisonian** [comment]
Updates look great - thanks @rollaball 🎉

**Pull request merged**